nixos/tests/gitea.nix

```diff
@@ -6,64 +6,99 @@
+      client = { config, pkgs, ... }: {
+        environment.systemPackages = [ pkgs.git ];
+      };
+      client2 = { config, pkgs, ... }: {
+        environment.systemPackages = [ pkgs.git ];
+      };
+    };
+
+    testScript = let
+      inherit (import ./ssh-keys.nix pkgs) snakeOilPrivateKey snakeOilPublicKey;
+    in ''
+      GIT_SSH_COMMAND = "ssh -i $HOME/.ssh/privk -o StrictHostKeyChecking=no"
+      REPO = "gitea@gitea_server:test/repo"
+      PRIVK = "${snakeOilPrivateKey}"
 
-    testScript = ''
       start_all()
 
-      machine.wait_for_unit("gitea.service")
-      machine.wait_for_open_port(3000)
-      machine.succeed("curl --fail http://localhost:3000/")
-    '';
-  };
+      client.succeed("mkdir /tmp/repo")
+      client.succeed("mkdir -p $HOME/.ssh")
+      client.succeed(f"cat {PRIVK} > $HOME/.ssh/privk")
+      client.succeed("chmod 0400 $HOME/.ssh/privk")
+      client.succeed("git -C /tmp/repo init")
+      client.succeed("echo hello world > /tmp/repo/testfile")
+      client.succeed("git -C /tmp/repo add .")
+      client.succeed("git config --global user.email test@localhost")
+      client.succeed("git config --global user.name test")
+      client.succeed("git -C /tmp/repo commit -m 'Initial import'")
+      client.succeed(f"git -C /tmp/repo remote add origin {REPO}")
 
-  postgres = makeTest {
-    name = "gitea-postgres";
-    meta.maintainers = [ maintainers.aanderse ];
+      gitea_server.wait_for_unit("gitea.service")
+      gitea_server.wait_for_open_port(3000)
+      gitea_server.succeed("curl --fail http://localhost:3000/")
 
-    machine =
-      { config, pkgs, ... }:
-      { services.gitea.enable = true;
-        services.gitea.database.type = "postgres";
-      };
+      gitea_server.succeed(
+          "curl --fail http://localhost:3000/user/sign_up | grep 'Registration is disabled. "
+          + "Please contact your site administrator.'"
+      )
+      gitea_server.succeed(
+          "su -l gitea -c 'GITEA_WORK_DIR=/var/lib/gitea gitea admin create-user "
+          + "--username test --password totallysafe --email test@localhost'"
+      )
 
-    testScript = ''
-      start_all()
+      api_token = gitea_server.succeed(
+          "curl --fail -X POST http://test:totallysafe@localhost:3000/api/v1/users/test/tokens "
+          + "-H 'Accept: application/json' -H 'Content-Type: application/json' -d "
+          + "'{\"name\":\"token\"}' | jq '.sha1' | xargs echo -n"
+      )
 
-      machine.wait_for_unit("gitea.service")
-      machine.wait_for_open_port(3000)
-      machine.succeed("curl --fail http://localhost:3000/")
-    '';
-  };
+      gitea_server.succeed(
+          "curl --fail -X POST http://localhost:3000/api/v1/user/repos "
+          + "-H 'Accept: application/json' -H 'Content-Type: application/json' "
+          + f"-H 'Authorization: token {api_token}'"
+          + ' -d \'{"auto_init":false, "description":"string", "license":"mit", "name":"repo", "private":false}\'''
+      )
```

Comment:
Maybe it's an idea to check if the repo exists from one of the clients? May be redundant, but since you only call the api from the gitea server itself it may be an idea.

Reply:
I clone this repository from `client2`, isn't that sufficient?

Reply:
It should be, yes. Just an idea.

Reply:
I see :) I added a simple test for this as well.